x-pack/plugins/apm/public/components/shared/critical_path_flamegraph/index.tsx

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { Chart, Datum, Flame, Settings } from '@elastic/charts';
+import {
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiLoadingSpinner,
+  euiPaletteColorBlind,
+} from '@elastic/eui';
+import { css } from '@emotion/css';
+import { useChartTheme } from '@kbn/observability-plugin/public';
+import { uniqueId } from 'lodash';
+import React, { useMemo, useRef } from 'react';
+import { FETCH_STATUS, useFetcher } from '../../../hooks/use_fetcher';
+import { CriticalPathFlamegraphTooltip } from './critical_path_flamegraph_tooltip';
+import { criticalPathToFlamegraph } from './critical_path_to_flamegraph';
+
+const chartClassName = css`
+  flex-grow: 1;
+`;
+
+export function CriticalPathFlamegraph(
+  props: {
+    start: string;
+    end: string;
+    traceIds: string[];
+    traceIdsFetchStatus: FETCH_STATUS;
+  } & ({ serviceName: string; transactionName: string } | {})
+) {
+  const { start, end, traceIds, traceIdsFetchStatus } = props;
+
+  const serviceName = 'serviceName' in props ? props.serviceName : null;
+  const transactionName =
+    'transactionName' in props ? props.transactionName : null;
+
+  // Use a reference to time range, to not invalidate the API fetch
+  // we only care for traceIds, start/end are there to limit the search
+  // request to a certain time range. It shouldn't affect the actual results
+  // of the search.
+  const timerange = useRef({ start, end });
+  timerange.current = { start, end };
+
+  const {
+    data: { criticalPath } = { criticalPath: null },
+    status: criticalPathFetchStatus,
+  } = useFetcher(
+    (callApmApi) => {
+      if (!traceIds.length) {
+        return Promise.resolve({ criticalPath: null });
+      }
+
+      return callApmApi('POST /internal/apm/traces/aggregated_critical_path', {
+        params: {
+          body: {
+            start: timerange.current.start,
+            end: timerange.current.end,
+            traceIds,
+            serviceName,
+            transactionName,
+          },
+        },
+      });
+    },
+    [timerange, traceIds, serviceName, transactionName]
+  );
+
+  const chartTheme = useChartTheme();
+
+  const isLoading =
+    traceIdsFetchStatus === FETCH_STATUS.NOT_INITIATED ||
+    traceIdsFetchStatus === FETCH_STATUS.LOADING ||
+    criticalPathFetchStatus === FETCH_STATUS.NOT_INITIATED ||
+    criticalPathFetchStatus === FETCH_STATUS.LOADING;
```

Comment:
We often check for both `NOT_INITIATED ` and `LOADING`. Perhaps we should have a helper in `use_fetcher` for this:

```suggestion
    isLoading(traceIdsFetchStatus) || isLoading(criticalPathFetchStatus)
```

Reply:
Yeah, in some cases we do need to differentiate though. Maybe there's a better word than `loading`, which implies data is being loaded in the background? `unresolved`?